odl/solvers/advanced/douglas_rachford.py

```diff
@@ -0,0 +1,152 @@
+        The Operator should be the proximal of ``f``.
+    prox_cc_g : `sequence` of `callable`'s
+        Sequence of functions returning an operator when called with step size.
+        The `Operator` should be the proximal of ``g_i^*``.
+    L : `sequence` of `Operator`'s
+        Sequence of `Opeartor`s` with as many elements as ``prox_cc_gs``.
+    tau : `float`
+        Step size for ``f``.
+    sigma : `sequence` of  `float`
+        Step size for the ``g_i``'s.
+    lam : `float`
+        Step size.
+    niter : `int`
+        Number of iterations.
+    callback : `Callback`, optional
+        Show partial results.
+
+    Other Parameters
+    ----------------
+    prox_cc_l : `sequence` of `callable`'s
+        Sequence of functions returning an operator when called with step size.
+        The `Operator` should be the proximal of ``l_i^*``.
+
+    Notes
+    -----
+    To guarantee convergence, the parameters ``tau``, ``sigma`` and ``L`` need
+    to satisfy
+
+    .. math::
+
+       \\tau \sum_{i=1}^n \\sigma_i ||L_i||^2 < 4
+
+    References
+    ----------
+    For references on the Forward-Backward algorithm, see [BH2013]_.
+    """
+
+    # Problem size
+    m = len(L)
+    if not all(isinstance(op, Operator) for op in L):
+        raise ValueError('`L` not a sequence of operators')
+    if not all(op.is_linear for op in L):
+        raise ValueError('not all operators in `L` are linear')
+    if not all(x in op.domain for op in L):
+        raise ValueError('`x` not in the domain of all operators')
+    if len(sigma) != m:
+        raise ValueError('len(sigma) != len(L)')
+    if len(prox_cc_g) != m:
+        raise ValueError('len(prox_cc_g) != len(L)')
+
+    # Get parameters
+    prox_cc_l = kwargs.pop('prox_cc_l', None)
+    if prox_cc_l is not None and len(prox_cc_l) != m:
+        raise ValueError('`prox_cc_l` not same length as `L`')
+
+    # Check for unused parameters
+    if kwargs:
+        raise TypeError('unexpected keyword argument: {}'.format(kwargs))
+
+    # Pre-allocate values
+    v = [Li.range.zero() for Li in L]
+
+    p1 = x.space.zero()
+    p2 = [Li.range.zero() for Li in L]
+
+    z1 = x.space.zero()
+    z2 = [Li.domain.zero() for Li in L]
+
+    w1 = x.space.zero()
+    w2 = [Li.range.zero() for Li in L]
+
+    for k in range(niter):
+        tmp_1 = sum(Li.adjoint(vi) for Li, vi in zip(L, v))
+        prox_f(tau)(x - (tau / 2.0) * tmp_1, out=p1)
+        w1.lincomb(2.0, p1, -1, x)
+
+        for i in range(m):
+            prox_cc_g[i](sigma[i])(v[i] + (sigma[i] / 2.0) * L[i](w1),
+                                   out=p2[i])
+            w2[i].lincomb(2.0, p2[i], -1, v[i])
+
+        tmp_2 = sum(Li.adjoint(wi) for Li, wi in zip(L, w2))
+        z1.lincomb(1.0, w1, - (tau / 2.0), tmp_2)
+        x += lam * (z1 - p1)
+
+        for i in range(m):
+            tmp = w2[i] + (sigma[i] / 2.0) * L[i](2.0 * z1 - w1)
+            if prox_cc_l is not None:
```

Comment:
We could allow for `None` entries in the sequence to signal trivial proximal (identity). Not sure though. Handling it would be easy with the `normalized_scalar_param_list` helper.